core/src/main/java/org/pentaho/platform/dataaccess/datasource/api/MetadataService.java

```diff
@@ -108,7 +111,26 @@ public List<String> getMetadataDatasourceIds() {
   }
 
   public MetadataTempFilesListDto uploadMetadataFilesToTempDir( InputStream metadataFile,
-      List<InputStream> localeFileStreams, List<String> localeFileNames ) throws Exception {
+     FormDataContentDisposition schemaFileInfo, List<InputStream> localeFileStreams, List<String> localeFileNames ) throws Exception {
+
+    String fileNameMain = schemaFileInfo.getFileName();
+    ZipInputStream zis = null;
+    ByteArrayOutputStream xmi = null;
+    if ( fileNameMain.endsWith( ".zip" ) ) {
+      zis = new ZipInputStream( metadataFile );
+      ZipEntry ze;
+      while ( ( ze = zis.getNextEntry() ) != null ) {
+        if ( ze.getName().endsWith( ".xmi" ) ) {
+          IOUtils.copy( zis, xmi = new ByteArrayOutputStream() );
+          zis.closeEntry();
+          break;
+        }
+        zis.closeEntry();
+      }
+      if ( xmi != null ) {
+        metadataFile = new ByteArrayInputStream( xmi.toByteArray() );
+      }
+    }
 
     String fileName = uploadFile( metadataFile );
```

Comment:
**This is a suggestion you don't have to change your code. the overall goal is achieve 80% or more code coverage on all new code.**

To make testing easier, a slight refactor can help. 


1) wrap the logic of extracting a xmi file from the zip in protected/default access modifier `InputStream extractXmiFile(FormDataContentDisposition schemaFileInfo)`
2) then the new body of `uploadMetadataFilesToTempDir(InputStream, FormDataContentDisposition, List<InputStream>, List<String>)` would call the original `uploadMetadataFilesToTempDir( InputStream,
      List<InputStream> , List<String> )`

```
public MetadataTempFilesListDto uploadMetadataFilesToTempDir( InputStream metadataFile, 
     FormDataContentDisposition schemaFileInfo, List<InputStream> localeFileStreams, List<String> localeFileNames )  {
     InputStream xmiFromZip = extractXmiFile( schemaFileInfo);
     uploadMetadataFilesToTempDir( (xmiFromZip != null ? xmiFromZip : metadataFile), 
localeFileStreams, localeFileNames ); // use original function
}
```
3) then you just have to write unit tests for `extractXmiFile(FormDataContentDisposition schemaFileInfo)` and that the `uploadMetadataFilesToTempDir(InputStream, FormDataContentDisposition, List<InputStream>, List<String>)` handles zips of normal xmi files, but you don't necessarily write unit test for original `uploadMetadataFilesToTempDir( InputStream, List<InputStream> , List<String> )` since it's "legacy code". But the code base will be better off if you do.